### PR TITLE
Use pyproject.toml

### PR DESCRIPTION
**Pull request by jarrodmillman**
This uses `pyproject.toml` and removes `setup.py`, `.mypi.ini`, and `ruff.toml`. I replicated the previous behavior where `setup.py` was grabbing optional dependencies from the `requirements` files using the same system we used in scikit-image (https://github.com/scikit-image/scikit-image/pull/6726 and https://github.com/scikit-image/scikit-image/pull/6763) and numpydoc (https://github.com/numpy/numpydoc/pull/474).

`pyproject.toml` is autogenerated from the `requirements` and `tools/pyproject.toml.in` via `python tools/generate_pyproject.toml.py`. The `pre-commit` monitors everything to make sure things don't get out-of-sync.

### CODE REVIEW

File: tools/generate_pyproject.toml.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+try:
+    import rtoml as toml
+except ImportError:
+    print("Please install `rtoml` first: `pip install rtoml`")
+    sys.exit(1)
+
+
+if not os.path.isfile("pyproject.toml"):
+    print("Please run this script from the skimage repository root:")
```
Comment (dschult) [inline tools/generate_pyproject.toml.py:14]:
```suggestion
    print("Please run this script from the networkx repository root:")
```

**Review by dschult — COMMENTED** [review]
Nice to have this working! :}

Can you explain why we don't need wheel or setuptools anymore?  What is doing the work that replaces them?

Picky nits:
- some places use single quotes ', others use double ", some use both -> pyproject.toml.in lines ~50 and ~100.  Does it matter? That is, is toml like python in its treatment of " vs '?
- I found a skimage -> networkx conversion.

**Comment by jarrodmillman** [comment]
> Can you explain why we don't need wheel or setuptools anymore? What is doing the work that replaces them?

In `pyproject.toml` we have
```
[build-system]
build-backend = 'setuptools.build_meta'
requires = ['setuptools>=61.2']
```

There is more detail here https://learn.scientific-python.org/development/guides/packaging_classic/#pep-517518-support-high-priority.

Basically, pip 10+ will create a new virtual environment, install `setuptools>=61.2`, and then build a wheel. setuptools requires wheel, so it will be brought in as well.

If we use `--no-build-isolation`, we would need to handle the virtual environment and installs ourselves.

**Comment by jarrodmillman** [comment]
I think the above means we can simplify things in our CI more (perhaps). I haven't had time to look into it, but was planning to talk to Henry about it soon.

**Review by dschult — APPROVED** [review]
Nice!   Thanks!  i approve this PR

**Review by rossbar — APPROVED** [review]
LGTM, thanks @jarrodmillman for making this transition. I'm not an expert on PEP517/518 but from my limited understanding I think this makes sense!

File: tools/pyproject.toml.in [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+[tool.ruff.per-file-ignores]
+"__init__.py" = ['I']
+"examples/*.py" = ['I']
+"doc/*.py" = ['I']
+"tools/*.py" = ['I']
+
+[tool.mypy]
+ignore_missing_imports = true
+exclude = 'subgraphviews|reportviews'
```
[the remark this answers is not in the document]
Comment (jarrodmillman) [reply tools/pyproject.toml.in:151]: It didn't seem necessary.

[the remark this answers is not in the document]
Comment (jarrodmillman) [reply tools/pyproject.toml.in:151]: With `exclude = 'subgraphviews|reportviews'` I get
```
$ mypy -p networkx
networkx/algorithms/similarity.py:675: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:676: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:677: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:678: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 553 source files
```
With the `exclude = 'subgraphviews|reportviews*'` I get
```
$  mypy -p networkx
networkx/algorithms/similarity.py:675: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:676: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:677: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:678: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 553 source files
```
With `exclude = 'subgraphviews'`I get
```
$ mypy -p networkx
networkx/algorithms/similarity.py:675: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:676: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:677: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:678: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/classes/tests/test_reportviews.py:1103: error: Incompatible types in assignment (expression has type "type[DiDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1113: error: Incompatible types in assignment (expression has type "type[OutDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1163: error: Incompatible types in assignment (expression has type "type[InDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1213: error: Incompatible types in assignment (expression has type "type[MultiDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1263: error: Incompatible types in assignment (expression has type "type[DiMultiDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1273: error: Incompatible types in assignment (expression has type "type[OutMultiDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
networkx/classes/tests/test_reportviews.py:1323: error: Incompatible types in assignment (expression has type "type[InMultiDegreeView]", base class "TestDegreeView" defined the type as "type[DegreeView]")  [assignment]
Found 7 errors in 1 file (checked 555 source files)
```
With `exclude = 'subgraphviews|reportviews'` and this removed
```
[[tool.mypy.overrides]]
module = 'networkx.classes.reportviews'
ignore_errors = true
```
I get
```
$ mypy -p networkx
networkx/classes/reportviews.py:1319: error: Incompatible types in assignment (expression has type "type[OutMultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/classes/reportviews.py:1371: error: Incompatible types in assignment (expression has type "type[MultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/classes/reportviews.py:1397: error: Incompatible types in assignment (expression has type "type[InMultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/algorithms/similarity.py:675: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:676: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:677: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:678: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
```
With `exclude = 'subgraphviews|reportviews*'` and this removed
```
[[tool.mypy.overrides]]
module = 'networkx.classes.reportviews'
ignore_errors = true
```
I get
```
$ mypy -p networkx
networkx/classes/reportviews.py:1319: error: Incompatible types in assignment (expression has type "type[OutMultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/classes/reportviews.py:1371: error: Incompatible types in assignment (expression has type "type[MultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/classes/reportviews.py:1397: error: Incompatible types in assignment (expression has type "type[InMultiEdgeDataView]", base class "OutEdgeView" defined the type as "type[OutEdgeDataView]")  [assignment]
networkx/algorithms/similarity.py:675: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:676: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:677: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
networkx/algorithms/similarity.py:678: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Found 3 errors in 1 file (checked 553 source files)
```

[the remark this answers is not in the document]
Comment (rossbar) [reply tools/pyproject.toml.in:151]: Seems like it doesn't make a difference then!

**Pull request merged**